### ISSUE TITLE
Field sampling near coast

### ISSUE THREAD

**Issue by pierrick-giffard** [OceanParcels/parcels#803]
I noticed that when particles are close to coast (I mean in the same cell that landmask, see first figure), field sampling is computed based on interpolation of oceanic true values but also based on zero values (because of land). Here for my example with temperature it results in particulary low temperatures near coast (second figure). A way to fix this might be to interpolate only with oceanic points available or with nearest neighbour ?

![Traj_temperature](https://user-images.githubusercontent.com/56306211/79207330-e1d41200-7e40-11ea-876a-4823d7db68a4.PNG)

![Temperature](https://user-images.githubusercontent.com/56306211/79207608-3081ac00-7e41-11ea-86c1-3c7cd2078a58.PNG)

**Comment by erikvansebille** [comment]
Thanks for reporting this Issue, @pierrick-giffard. This problem of interpolating near land indeed pops up quite frequently in my own research too,  and I don't have a perfect solution. I do have a few thoughts, though

1) The simplest 'solution' is to set the `interp_method` of your temperature field to `nearest`. You will then be able to remove all the locations where the particle temperature is zero in post-processing. However, this is not ideal because you don't go to the root of the cause (there will still be 'zeros', they're just easier to identify now) and you also have to use an interpolation method that you don't want to use * ...

2) Write a new interpolation method that 'ignores' zeros of land values. However, I don't think it's trivial what this interpolation method would look like, let alone implement it into [parcels.h](https://github.com/OceanParcels/parcels/blob/master/parcels/include/parcels.h).

3) Change your `Field` data and change the zeros to e.g. the closest non-zero value. The tricky thing here is that you're probably using deferred-loading of the `FieldSet`, so you have to do this 'on the fly'. While that is in principle possible in Parcels (using the `FieldSet.compute_on_defer` method that is used e.g. [here](https://github.com/OceanParcels/parcels/blob/master/tests/test_fieldset.py#L665-L673)), this is not for the faint-at-heart. It would also likely come at a considerable efficiency loss

So in short, there is no 'perfect' solution, I think. Unless you have an equation in mind you would want to use for the interpolation, then we could try option 2)

_* note that nearest interpolation is the default tracer interpolation for C-grids by the way, see [@delandmeterp's paper](https://www.geosci-model-dev.net/12/3571/2019/gmd-12-3571-2019.html) for an explanation why_

**Comment by pierrick-giffard** [comment]
Thanks for your answer Erik. I was thinking it might be possible to implement something like this, supposing we know which of the 4 points around the parcel are on land (land could be determined by abs(data) < 1e-14 ? ).

- If the **4 points** are ocean: compute bilinear interpolation as usual.

- If only **1 point** is ocean: take the value of this point.

- If only **2 points** are ocean: project the parcel on the line drawn by the 2 points and compute 1D interpolation. In case this line is the diagonal, I found that the distance between the upper left corner and the projected parcel initially located at (eta, xsi) is w = sqrt(2)/2 * (1 + xsi - eta). 
The 1D interpolation will then be T = (1-w) * T[0,1] + w * T[1,0].

- If **3 points** are ocean, there are 2 cases:
     1. The parcel is outside the triangle made by the 3 points. Then project the parcel on the line drawn by the two diagonal points and compute 1D interpolation.
     2. The parcel is inside the triangle made by the 3 points. Then compute interpolation in the triangle based on barycentric coordinates (https://math.stackexchange.com/questions/1099390/billinear-interpolation-of-3-points).
    The temperature is then:
        T = (1-xsi-eta) * T[0,0] + xsi * T[1,0] + eta * T[0,1]  if bottom left corner is ocean
    or T = (1-xsi-eta) * T[1,1] + xsi * T[1,0] + eta * T[0,1] if upper right corner is ocean

What do you think about implementing something like this? I hope I was clear enough...

**Comment by erikvansebille** [comment]
Thanks @pierrick-giffard. These are interesting ideas. Could you try putting these into `parcels.h` yourself? You could for example adapt https://github.com/OceanParcels/parcels/blob/master/parcels/include/parcels.h#L93-L103. If you create a Pull Request when you're ready, we'll do a review of your code and merge it

Note, by the way, that your solution with four points is for 2D fields only. There's of course also trilinear interpolation, which involved 8 points. Perhaps best to start with the 2D formulation, but keep 3D in mind too

**Comment by pierrick-giffard** [comment]
I have been investigating a bit deeper different possibilities for interpolation near coast and it seems that the one I suggested above is a bit too complex, especially in 3D. However, I found that the method called ` Inverse distance weighted interpolation`  (it seems to be well known) is really close to linear solutions in our case, given that fields gradients aren't too high in oceanography (see figure below). It attributes to each point the following weight: `weight(i) = 1/distance(i) / sum(1/distance(i))`

Obviously, in case all points are in the ocean we would keep the bilinear method. In case the particle is exactly at a point location (distance(i) = 0), it would return the field value at this point.

![image](https://user-images.githubusercontent.com/56306211/79886381-1162bb80-83f9-11ea-8fb3-943a4e5c9918.png)
If this method looks good for you, I will create a pull request with the C code.

**Comment by erikvansebille** [comment]
Thanks @pierrick-giffard. This does look like a very interesting option for interpolation yes. It would also be fairly straight-forward to extend this to 3D, right? 

Would you do this in the unit cell ([Fig 3b of @delandmeterp's paper](https://www.geosci-model-dev.net/12/3571/2019/gmd-12-3571-2019.pdf)), so in \xsi, \eta, \zeta coordinates?

I'd say yes, please go ahead and try implement this into the C code; let me know if we can help!

**Comment by pierrick-giffard** [comment]
Yes, it is fairly straight-forward to extend this to 3D. The distance will be calculated as `sqrt(xsi² + eta² + zeta²)` for the point origin.

**Comment by pierrick-giffard** [comment]
One question, the best way to determine whether a point is on land or not is to evaluate if `abs(data) < 1e-14` ?

**Comment by pierrick-giffard** [comment]
An other question, at this point can a particle be completely surrounded by land or the ErrorCode will be triggered before?

**Comment by erikvansebille** [comment]
> One question, the best way to determine whether a point is on land or not is to evaluate if abs(data) < 1e-14 ?

Yes, that would be a good way to test if Field is zero

> An other question, at this point can a particle be completely surrounded by land or the ErrorCode will be triggered before?

We do not throw ErrorCodes for being on land, so it could indeed be possible that a Particle is completely surrounded by land points. Perhaps an idea to throw an Error in this Kernel then? If you just raise the Error, then I will implement all the backend for it

**Comment by pierrick-giffard** [comment]
Using  `interp_method='linear_invdist_land_tracer'` fixes this issue. I don't notice anymore temperature decrease near coast.
